--- a/505.mcf_r/spec_qsort/spec_qsort.c
+++ b/505.mcf_r/spec_qsort/spec_qsort.c
@@ -113,0 +114,5 @@
+/* Add instrumentation function for the pintool */
+#ifdef PIN_ROI_INSTRUMENTATION
+#include <pinroi.h>
+#endif
+
@@ -116,0 +122,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+	custom_roi_begin("spec_qsort");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 1");
+#endif
@@ -130,0 +142,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+		custom_roi_end("spec_qsort");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -178,0 +196,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("spec_qsort");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
