--- a/520.omnetpp_r/simulator/cgate.cc
+++ b/520.omnetpp_r/simulator/cgate.cc
@@ -319,0 +320,5 @@
+/* Add instrumentation function for the pintool */
+#ifdef PIN_ROI_INSTRUMENTATION
+#include <pinroi.h>
+#endif
+
@@ -321,0 +327,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_begin("cGate::deliver");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 3");
+#endif
@@ -324,0 +336,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("cGate::deliver");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 3");
+#endif
@@ -334,0 +352,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("cGate::deliver");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 3");
+#endif
@@ -339,0 +363,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("cGate::deliver");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 3");
+#endif
