--- a/520.omnetpp_r/simulator/cmessageheap.cc
+++ b/520.omnetpp_r/simulator/cmessageheap.cc
@@ -223 +223,6 @@
-cMessage *cMessageHeap::removeFirst()
+/* Add instrumentation function for the pintool */
+#ifdef PIN_ROI_INSTRUMENTATION
+#include <pinroi.h>
+#endif
+
+cMessage *cMessageHeap::removeFirst() __attribute__((noinline)) /* Disable inlining for ZRay ROI */
@@ -224,0 +230,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_begin("cMessageHeap::removeFirst");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 1");
+#endif
@@ -232,0 +244,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("cMessageHeap::removeFirst");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -234,0 +252,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("cMessageHeap::removeFirst");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
