--- a/525.x264_r/x264_src/common/mc.c
+++ b/525.x264_r/x264_src/common/mc.c
@@ -123,0 +124,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_begin("mc_weight");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 1");
+#endif
@@ -135,0 +142,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("mc_weight");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -231,0 +244,5 @@
+/* Add instrumentation function for the pintool */
+#ifdef PIN_ROI_INSTRUMENTATION
+#include <pinroi.h>
+#endif
+
@@ -236,0 +254,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_begin("get_ref");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 1");
+#endif
@@ -247,0 +271,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("get_ref");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -252,0 +282,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("get_ref");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -257,0 +293,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("get_ref");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
