--- a/525.x264_r/x264_src/common/pixel.c
+++ b/525.x264_r/x264_src/common/pixel.c
@@ -234 +234,6 @@
-static NOINLINE int x264_pixel_satd_8x4( uint8_t *pix1, int i_pix1, uint8_t *pix2, int i_pix2 )
+/* Add instrumentation function for the pintool */
+#ifdef PIN_ROI_INSTRUMENTATION
+#include <pinroi.h>
+#endif
+
+static NOINLINE int x264_pixel_satd_8x4( uint8_t *pix1, int i_pix1, uint8_t *pix2, int i_pix2 ) __attribute__((noinline)) /* Disable inlining for ZRay ROI */
@@ -235,0 +241,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_begin("x264_pixel_satd_8x4");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 1");
+#endif
@@ -251,0 +263,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("x264_pixel_satd_8x4");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
