--- a/511.povray_r/objects.cpp
+++ b/511.povray_r/objects.cpp
@@ -39,0 +40,5 @@
+/* Add instrumentation function for the pintool */
+#ifdef PIN_ROI_INSTRUMENTATION
+#include <pinroi.h>
+#endif
+
@@ -207,0 +213,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_begin("Ray_In_Bound");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 1");
+#endif
@@ -218,0 +230,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("Ray_In_Bound");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -225,0 +243,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("Ray_In_Bound");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
