--- a/531.deepsjeng_r/search.cpp
+++ b/531.deepsjeng_r/search.cpp
@@ -662,0 +663,5 @@
+/* Add instrumentation function for the pintool */
+#ifdef PIN_ROI_INSTRUMENTATION
+#include <pinroi.h>
+#endif
+
@@ -666,0 +672,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_begin("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 1");
+#endif
@@ -690,0 +702,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -696,0 +714,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -700,0 +724,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -711,0 +741,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -723,0 +759,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -729,0 +771,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -733,0 +781,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -739,0 +793,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -764,0 +824,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -767,0 +833,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -773 +844,7 @@
-                return escore;
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif          
+		return escore;
@@ -803,0 +881,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -836,0 +920,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -841,0 +931,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -850,0 +946,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -855,0 +957,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -970,0 +1078,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -1254,0 +1368,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -1267,0 +1387,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -1288,0 +1414,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -1291,0 +1423,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -1295,0 +1433,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -1303,0 +1447,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("search");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
