--- a/500.perlbench_r/pp_ctl.c
+++ b/500.perlbench_r/pp_ctl.c
@@ -40,0 +41,5 @@
+/* Add instrumentation function for the pintool */
+#ifdef PIN_ROI_INSTRUMENTATION
+#include <pinroi.h>
+#endif
+
@@ -2065,0 +2071,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_begin("pp_enter");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 1");
+#endif
@@ -2073,0 +2085,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("pp_enter");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
