--- a/500.perlbench_r/sv.c
+++ b/500.perlbench_r/sv.c
@@ -4292,0 +4293,5 @@
+/* Add instrumentation function for the pintool */
+#ifdef PIN_ROI_INSTRUMENTATION
+#include <pinroi.h>
+#endif
+
@@ -4295,0 +4301,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_begin("Perl_sv_setsv_flags");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 1");
+#endif
@@ -4302 +4313,7 @@
-    if (UNLIKELY( sstr == dstr ))
+    if (UNLIKELY( sstr == dstr )) {
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("Perl_sv_setsv_flags");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -4303,0 +4321 @@
+    }
@@ -4325,0 +4344,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("Perl_sv_setsv_flags");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -4356,0 +4381,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("Perl_sv_setsv_flags");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -4386,0 +4417,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("Perl_sv_setsv_flags");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -4439,0 +4476,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("Perl_sv_setsv_flags");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -4491,0 +4534,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("Perl_sv_setsv_flags");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -4494,0 +4543,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("Perl_sv_setsv_flags");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -4500,0 +4555,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("Perl_sv_setsv_flags");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -4780,0 +4841,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("Perl_sv_setsv_flags");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
