--- a/557.xz_r/liblzma/lzma/lzma_decoder.c
+++ b/557.xz_r/liblzma/lzma/lzma_decoder.c
@@ -281,0 +282,4 @@
+/* Add instrumentation function for the pintool */
+#ifdef PIN_ROI_INSTRUMENTATION
+#include <pinroi.h>
+#endif
@@ -287,0 +292,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+	custom_roi_begin("lzma_decode");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 1");
+#endif
@@ -293 +303,9 @@
-	if (!rc_read_init(&coder->rc, in, in_pos, in_size))
+	// if (!rc_read_init(&coder->rc, in, in_pos, in_size))
+	// 	return LZMA_OK;
+	if (!rc_read_init(&coder->rc, in, in_pos, in_size)) {
+#ifdef PIN_ROI_INSTRUMENTATION
+		custom_roi_end("lzma_decode");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
@@ -294,0 +313 @@
+	}
@@ -833,0 +853,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+	custom_roi_end("lzma_decode");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
