--- a/500.perlbench_r/cpan/Digest-SHA/src/sha64bit.c
+++ b/500.perlbench_r/cpan/Digest-SHA/src/sha64bit.c
@@ -94,0 +95,4 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+#include <pinroi.h>
+#endif
+
@@ -96,0 +101,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_begin("sha512");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_BEGIN 1");
+#endif
@@ -114,0 +125,6 @@
+#ifdef PIN_ROI_INSTRUMENTATION
+    custom_roi_end("sha512");
+#endif
+#ifdef ZRAY_ROI_INSTRUMENTATION
+asm volatile("#ZRAY_ROI_END 1");
+#endif
